src/Microsoft.JSInterop/JSRuntimeBase.cs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Abstract base class for a JavaScript runtime.
+    /// </summary>
+    public abstract class JSRuntimeBase : IJSRuntime
+    {
+        private long _nextPendingTaskId = 1; // Start at 1 because zero signals "no response needed"
+        private readonly ConcurrentDictionary<long, object> _pendingTasks
+            = new ConcurrentDictionary<long, object>();
+
+        /// <summary>
+        /// Invokes the specified JavaScript function asynchronously.
+        /// </summary>
+        /// <typeparam name="T">The JSON-serializable return type.</typeparam>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <code>"someScope.someFunction"</code> will invoke the function <code>window.someScope.someFunction</code>.</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>An instance of <typeparamref name="T"/> obtained by JSON-deserializing the return value.</returns>
+        public Task<T> InvokeAsync<T>(string identifier, params object[] args)
+        {
+            // TODO: Auto-timeout
+
+            var taskId = Interlocked.Increment(ref _nextPendingTaskId);
+            var tcs = new TaskCompletionSource<T>();
+            _pendingTasks[taskId] = tcs;
+
+            try
+            {
+                BeginInvokeJS(taskId, identifier, args?.Length > 0 ? Json.Serialize(args) : null);
+                return tcs.Task;
+            }
+            catch
+            {
+                _pendingTasks.TryRemove(taskId, out _);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Begins an asynchronous function invocation.
+        /// </summary>
+        /// <param name="asyncHandle">The identifier for the function invocation, or zero if no async callback is required.</param>
+        /// <param name="identifier">The identifier for the function to invoke.</param>
+        /// <param name="argsJson">A JSON representation of the arguments.</param>
+        protected abstract void BeginInvokeJS(long asyncHandle, string identifier, string argsJson);
+
+        internal void EndInvokeDotNet(string callId, bool success, object resultOrExceptionMessage)
```

Comment:
With these begin-end pattern things, it's possible to have a bug where begin/end are called on different runtime instances - (if anything changes the instance of the runtime during execution). I don't think it's likely that anyone will do that - but just in case.

Reply:
It is possible, that's true. But our work on sync context should make it easier to get this right than wrong in typical cases. In the even that a developer somehow still gets it wrong, it should be pretty detectable because there will be a lot of errors of the form "there is no pending task with ID x". If you have an idea for how we should improve this please let me know :)